### ISSUE TITLE
TaxonomyFormat assumes first line is header

### ISSUE THREAD

**Issue by jairideout** [qiime2/q2-types#107]
When reading `TaxonomyFormat` with any of the transformers, the first line is assumed to be a (non-comment) header, followed by the taxonomy mapping lines. The sniffer is very lenient and only cares that the file is two-column TSV.

Not all taxonomy files include a header (for example, Greengenes). When a transformer is invoked to read the file, the first line is interpreted as a header, causing the feature ID to be set as `Index.name` and the taxonomy string to be set as `Series.name`.

For example, suppose we have the following `taxonomy.tsv` file (I used the first few lines from the Greengenes taxonomy map):

```
228054  k__Bacteria; p__Cyanobacteria; c__Synechococcophycideae; o__Synechococcales; f__Synechococcaceae; g__Synechococcus; s__
228057  k__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rickettsiales; f__Pelagibacteraceae; g__; s__
73627   k__Bacteria; p__Actinobacteria; c__Actinobacteria; o__Actinomycetales; f__Mycobacteriaceae; g__Mycobacterium; s__
378462  k__Bacteria; p__Firmicutes; c__Bacilli; o__Bacillales; f__Staphylococcaceae; g__Staphylococcus; s__
89370   k__Bacteria; p__Firmicutes; c__Bacilli; o__Bacillales; f__Bacillaceae; g__Anoxybacillus; s__kestanbolensis
```

Reading this file into a `pd.Series` (all transformers are affected, it's not limited to the `pd.Series` transformer):

```python
In [1]: from qiime2.plugin.util import transform

In [2]: from q2_types.feature_data import TaxonomyFormat

In [3]: import pandas as pd

In [4]: taxonomy_series = transform('taxonomy.tsv', from_type=TaxonomyFormat, to_type=pd.Series)

In [5]: taxonomy_series
Out[5]:
228054
228057    k__Bacteria; p__Proteobacteria; c__Alphaproteo...
73627     k__Bacteria; p__Actinobacteria; c__Actinobacte...
378462    k__Bacteria; p__Firmicutes; c__Bacilli; o__Bac...
89370     k__Bacteria; p__Firmicutes; c__Bacilli; o__Bac...
Name: k__Bacteria; p__Cyanobacteria; c__Synechococcophycideae; o__Synechococcales; f__Synechococcaceae; g__Synechococcus; s__, dtype: object

In [6]: # :(

In [7]:
```

The Series has its `Index.name` set to "228054" and its `Series.name` to "k__Bacteria; p__Cyanobacteria; c__Synechococcophycideae; o__Synechococcales; f__Synechococcaceae; g__Synechococcus; s__".

During classification, if a query sequence is assigned to the first reference sequence taxonomy (i.e. the one that's misinterpreted as a header), I *think* the code will error with an `IndexError` coming from pandas. This happened with @nbokulich's `vsearch` classifier (not yet in master), and adding a header line to the Greengenes file appears to fix the issue (he's not receiving the error anymore at least).

I don't think we've seen this error with the existing classifiers because we've never had a query sequence assigned to the first reference sequence (e.g. just by chance). I suspect (and hope) that the code would fail in a similar way with an `IndexError`, but haven't confirmed.

I propose that we require `TaxonomyFormat` (both when sniffing and reading) to have the following header (we've been using this header in the unit tests, maybe elsewhere):

```
Feature ID<tab>Taxon
```

... optionally followed by other columns that are ignored. We could finesse the column names a little -- something like `feature_id` and `taxon` would be easier to access from pandas objects. This is a minor detail we can work out later.

If we go with a stricter format, such as what I'm proposing, then importing files without the appropriate header (e.g. Greengenes and other reference databases) will raise an error and the header will have to be added to the file in order to import. This is annoying, but I really think we should stop supporting tabular files without headers, especially because we want the .qza file formats to be as self-documenting as possible.

Thoughts? cc: @gregcaporaso, @nbokulich, @BenKaehler, @ebolyen, @thermokarst, @jakereps 

Thanks @nbokulich for finding and reporting this bug!

**Comment by ebolyen** [comment]
:+1: I think we are using `Index` (which may also be a good name) somewhere else, but not sure where...

Also, I agree that this is just going to be a catastrophic failure, so I don't think we need to worry about incorrect results at least.

**Comment by jairideout** [comment]
> Also, I agree that this is just going to be a catastrophic failure, so I don't think we need to worry about incorrect results at least.

I think it's likely to error noisily, but it'd be good to confirm that the error is produced as expected, and that the reason we haven't seen this error yet is due to non-assignment of the first ref seq. I only tested this with @nbokulich's branch on an entirely new method. With Greengenes having integer IDs, coupled with pandas `__getitem__` indexing instead of `.loc` or `.iloc`, there's a lot that can go wrong (e.g. using an integer ID accidentally as an index, and the indexing is off-by-one since the header is misinterpreted).

**Comment by ebolyen** [comment]
> With Greengenes having integer IDs, coupled with pandas __getitem__ indexing instead of .loc or .iloc, there's a lot that can go wrong 

I hadn't even put two and two together, that is a really good point. We *need* to figure out metadata and indexes.

**Comment by gregcaporaso** [comment]
Thanks @nbokulich and @jairideout. I agree about requiring a header line in tsv files.

**Comment by jairideout** [comment]
> I hadn't even put two and two together, that is a really good point. We need to figure out metadata and indexes.

I totally agree -- this [came up yesterday](https://github.com/qiime2/q2-diversity/issues/91) in a different plugin, and it required the user to rerun their entire analysis, ouch!).

**Comment by gregcaporaso** [comment]
>> I hadn't even put two and two together, that is a really good point. We need to figure out metadata and indexes.

> I totally agree 

Should we try to discuss this on Tuesday to start working on a plan?

**Comment by jairideout** [comment]
That'd be great! I think CI/CD and Metadata/Indexes are top priority.

**Comment by BenKaehler** [comment]
I just rediscovered this bug. The `scikit-learn` based classifiers are quite permissive, so this just results in the top line being missing from the training set, and that just became an issue for me.

On whether the input files should be required to have headings, looking at the greengenes and unite data sets that I have, that doesn't seem to be the convention. That doesn't mean that they shouldn't be stored with headings in the `.qza`s, just that I don't reckon it's what people would be expecting when they try to load up standard data sets.

**Comment by jairideout** [comment]
Thanks @BenKaehler! We need to get this fixed for the next release (2017.5). I'm happy to help with that.

> On whether the input files should be required to have headings, looking at the greengenes and unite data sets that I have, that doesn't seem to be the convention. That doesn't mean that they shouldn't be stored with headings in the .qzas, just that I don't reckon it's what people would be expecting when they try to load up standard data sets.

I suppose we could support two taxonomy formats-- one with a *standardized header* (this one would also be the .qza's internal format) and one without a header (to support Greengenes, etc.). What do you guys think?

cc: @gregcaporaso @ebolyen @thermokarst @nbokulich

**Comment by nbokulich** [comment]
Supporting import of headless or headed taxonomy files sounds useful. maybe add a boolean parameter `headless` to specify whether the input file is headless or not? If headless, append the expected header for that semantic type. One advantage of doing this is that it is flexible in case this issue may arise with other semantic types. 

Either that or at least raise an error if the expected header is absent — I prefer @jairideout 's suggestion though.

**Comment by jairideout** [comment]
> maybe add a boolean parameter headless to specify whether the input file is headless or not? 

👍 The framework doesn't currently support parameters to file formats / directory formats but there's been discussion in the past to support this feature.

With the current framework support, if we have two file formats (header & headerless), users would need to supply an explicit `--source-format` when importing the headerless file format (since the header file format would be the default format used in .qza's).

So if we define two file formats, users will still need to manually intervene when importing the headerless format. If there is a single format, users would need to intervene by adding the appropriate header to the file before importing.

Having two formats sounds a little more user-friendly but in either case the user will have to take some sort of action when supplying a headerless format (there isn't a "seamless" solution possible right now).

I don't have strong feelings either way (and happy to implement whatever we decide on).

**Comment by ebolyen** [comment]
I would definitely lean towards the two formats (and in the future collapsing to a parameter [does this potentially bump priority?]). Having users edit their input can end poorly sometimes.
